lib/puma/server.rb

```diff
@@ -326,6 +317,20 @@ def handle_servers
       @events.fire :state, :done
     end
 
+    def accept_socket(sock)
+      sock.accept_nonblock
+    rescue SystemCallError
+      # nothing
+    rescue Errno::ECONNABORTED
+      # client closed the socket even before accept
+      begin
+        sock.close
+      rescue
+        Thread.current.purge_interrupt_queue if Thread.current.respond_to? :purge_interrupt_queue
+      end
+    end
+    private :accept_socket
+
```

Comment:
Extracted this logic into a private method both to simplify the accept-loop, but more importantly to allow the to be stubbed in the `test_blocks_new_connection` unit-test.

Reply:
hmm, the original `ECONNABORTED` rescue clause had `io.close` not `sock.close`, I should change this back to the original behavior.

That  said, the line is _ancient_, dating back to 2006 (https://github.com/puma/puma/commit/ab3c8082de82e6fc96838d444be06432620743ab) (!) and never covered by any tests, so it's very likely this is 'superstitious' code that should probably be changed/removed separately anyway. (In my reading of the current code, `ECONNABORTED` would only ever be thrown by `accept_nonblock` if the listener socket is closed, but in that case `io` would be `nil` causing `io.close` to always fail..)

Reply:
I managed to figure out how to stub the `accept` call without needing to extract the method, so I'll move this refactoring/cleanup to a separate PR.